### PR TITLE
posix: conditionally compile perror pthread_common and nanosleep

### LINKED ISSUE #50654: Some files are being ALWAYS built, without them being used
**Describe the bug**
Many files are being by default although they are not being used pulling files from upper layers and other subsystems.

**To Reproduce**
1. `west build -b qemu_x86 samples/synchronization/`
2. `cat  build/compile_commands.json  | jq .[].file   |less`
3. Review list of files being built

**Expected behavior**
We should only build files that are relevant to the application being built

**Impact**
 We should not be pulling upper layer code into the core kernel and basic needed set creating unnecessary dependenies.

**Logs and console output**
example of files being built for EACH application, whether needed or not, maybe this more, but those I can spot immediately. 

```
"/home/nashif/zephyrproject/zephyr/lib/posix/perror.c"
"/home/nashif/zephyrproject/zephyr/lib/posix/pthread_common.c"
"/home/nashif/zephyrproject/zephyr/lib/posix/nanosleep.c"
"/home/nashif/zephyrproject/zephyr/lib/os/crc32c_sw.c"
"/home/nashif/zephyrproject/zephyr/lib/os/crc32_sw.c"
"/home/nashif/zephyrproject/zephyr/lib/os/crc16_sw.c"
"/home/nashif/zephyrproject/zephyr/lib/os/crc8_sw.c"
"/home/nashif/zephyrproject/zephyr/lib/os/crc7_sw.c"
"/home/nashif/zephyrproject/zephyr/lib/os/dec.c"
"/home/nashif/zephyrproject/zephyr/lib/os/fdtable.c"
"/home/nashif/zephyrproject/zephyr/lib/os/hex.c"
"/home/nashif/zephyrproject/zephyr/lib/os/notify.c"
"/home/nashif/zephyrproject/zephyr/lib/os/onoff.c"
```

### PR DESCRIPTION
**Pull request by cfriedt**
Previously, these files were compiled unconditionally.

Partially fixes #50654

### CODE REVIEW

**Comment by nashif** [comment]
@cfriedt thank you for the prompt reaction :-D

**Review by carlescufi — DISMISSED** [review]

**Review by stephanosio — CHANGES_REQUESTED** [review]
There is now `No SOURCES given to Zephyr library: lib__posix` CMake error when no POSIX config is enabled.

**Comment by cfriedt** [comment]
> There is now `No SOURCES given to Zephyr library: lib__posix` CMake error when no POSIX config is enabled.

Let me get that fixed

File: lib/posix/CMakeLists.txt [hunk 1]
```diff
@@ -26,13 +27,12 @@ zephyr_library_sources_ifdef(CONFIG_POSIX_FS fs.c)
...
 zephyr_library_include_directories(
   ${ZEPHYR_BASE}/kernel/include
   ${ARCH_DIR}/${ARCH}/include
 )
 
-zephyr_library_link_libraries(posix_subsys)
+if (CONFIG_POSIX_API OR CONFIG_PTHREAD_IPC OR CONFIG_POSIX_CLOCK OR
+    CONFIG_POSIX_MQUEUE OR CONFIG_POSIX_FS OR CONFIG_EVENTFD OR CONFIG_GETOPT)
+  zephyr_library_link_libraries(posix_subsys)
+endif()
```
Comment (stephanosio) [inline lib/posix/CMakeLists.txt:38]: well, this is messy ... ideally, there would be a top-level POSIX subsystem symbol (e.g. `CONFIG_POSIX`), but that is not the case at the moment.

another option would be to just set `zephyr_library_property(ALLOW_EMPTY TRUE)`.

**Review by stephanosio — APPROVED** [review]

**Review by enjiamai — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**